File: tests/e2e/skill_test.go
package e2e

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
)

// Helpers for writing a config.toml with skills declared.

func writeConfigWithLocalSkill(t *testing.T, h *harness, skillName, skillDirPath string) {
	t.Helper()
	cfg := h.readFile(h.configFile())
	skillEntry := fmt.Sprintf("\n[skills.%s]\nsource = %q\n", skillName, skillDirPath)
	h.writeFile(h.configFile(), cfg+skillEntry)
}

func writeConfigWithGitSkill(t *testing.T, h *harness, skillName, repoURL, ref string) {
	t.Helper()
	cfg := h.readFile(h.configFile())
	skillEntry := fmt.Sprintf("\n[skills.%s]\nsource = %q\nref = %q\n", skillName, repoURL, ref)
	h.writeFile(h.configFile(), cfg+skillEntry)
}

// A local skill declared in config must appear as a symlink inside every
// enabled vendor's skills directory after sync.
func TestSkillSync_LocalSkill_AppearsInVendorSkillsDir(t *testing.T) {
	if isWindows() {
		t.Skip("symlink tests require Unix")
	}
	h := newHarness(t)
	h.bootstrap()

	skillFixtureDir := repoFixtureDir(t, "simple_skill")
	writeConfigWithLocalSkill(t, h, "simple-skill", skillFixtureDir)

	h.mustRun("sync")

	for vendor, skillsDir := range h.vendorSkillsDirs() {
		t.Run(vendor, func(t *testing.T) {
			skillPath := filepath.Join(skillsDir, "simple-skill")
			skillMD := filepath.Join(skillPath, "SKILL.md")
			got := h.readFile(skillMD)
			if !strings.Contains(got, "simple-skill") {
				t.Errorf("expected SKILL.md to contain skill name, got:\n%s", got)
			}
		})
	}
}

// Vendor skill entries must be symlinks pointing into the ponte store, not
// copies. This is the core immutability guarantee.
func TestSkillSync_LocalSkill_IsSymlinkedFromStore(t *testing.T) {
	if isWindows() {
		t.Skip("symlink tests require Unix")
	}
	h := newHarness(t)
	h.bootstrap()

	skillFixtureDir := repoFixtureDir(t, "simple_skill")
	writeConfigWithLocalSkill(t, h, "simple-skill", skillFixtureDir)

	h.mustRun("sync")

	h.assertIsStoreSymlink(h.vendorSkillPath("claude-code", "simple-skill"))
}

// Instruction files must also be symlinks into the store after sync.
func TestSkillSync_InstructionFile_IsSymlinkedFromStore(t *testing.T) {
	if isWindows() {
		t.Skip("symlink tests require Unix")
	}
	h := newHarness(t)
	h.bootstrap()

	h.assertIsStoreSymlink(h.vendorPaths()["claude-code"])
}

// Two syncs with identical inputs produce the same store generation hash:
// the symlink target does not change and no new store directory is created.
func TestSkillSync_SameInputs_ReuseStoreGeneration(t *testing.T) {
	if isWindows() {
		t.Skip("symlink tests require Unix")
	}
	h := newHarness(t)
	h.bootstrap()

	skillFixtureDir := repoFixtureDir(t, "simple_skill")
	writeConfigWithLocalSkill(t, h, "simple-skill", skillFixtureDir)
	h.mustRun("sync")

	firstTarget, err := os.Readlink(h.vendorPaths()["claude-code"])
	if err != nil {
		t.Fatalf("readlink: %v", err)
	}

	h.mustRun("sync")

	secondTarget, err := os.Readlink(h.vendorPaths()["claude-code"])
	if err != nil {
		t.Fatalf("readlink after second sync: %v", err)
	}
	if firstTarget != secondTarget {
		t.Errorf("expected same store generation on identical inputs\nfirst:  %s\nsecond: %s", firstTarget, secondTarget)
	}
}

// Adding a skill to the config changes the store generation: the symlink
// target moves to a new generation that contains the skill.
func TestSkillSync_AddingSkill_CreatesNewGeneration(t *testing.T) {
	if isWindows() {
		t.Skip("symlink tests require Unix")
	}
	h := newHarness(t)
	h.bootstrap()
	h.mustRun("sync")

	firstTarget, _ := os.Readlink(h.vendorPaths()["claude-code"])

	skillFixtureDir := repoFixtureDir(t, "simple_skill")
	writeConfigWithLocalSkill(t, h, "simple-skill", skillFixtureDir)
	h.mustRun("sync")

	secondTarget, _ := os.Readlink(h.vendorPaths()["claude-code"])
	if firstTarget == secondTarget {
		t.Error("expected a new store generation after adding a skill, but symlink target did not change")
	}

	// Skill must be present in the new generation.
	skillPath := h.vendorSkillPath("claude-code", "simple-skill")
	if _, err := os.Stat(skillPath); err != nil {
		t.Errorf("expected skill to exist at %s: %v", skillPath, err)
	}
}

// A git-backed skill is cloned, checked out at the given ref, and symlinked
// into vendor skill directories. Uses a local git repo as the remote to avoid
// network dependency.
func TestSkillSync_GitSkill_ClonesAndLinks(t *testing.T) {
	if isWindows() {
		t.Skip("symlink tests require Unix")
	}
	if _, err := exec.LookPath("git"); err != nil {
		t.Skip("git not available")
	}

	h := newHarness(t)
	h.bootstrap()

	repoPath, commitSHA := createLocalGitSkillRepo(t)

	writeConfigWithGitSkill(t, h, "git-skill", "file://"+repoPath, commitSHA)

	h.mustRun("sync")

	skillPath := h.vendorSkillPath("claude-code", "git-skill")
	skillMD := filepath.Join(skillPath, "SKILL.md")
	got := h.readFile(skillMD)
	if !strings.Contains(got, "git-skill") {
		t.Errorf("expected SKILL.md from git repo, got:\n%s", got)
	}
}

// createLocalGitSkillRepo initialises a bare git repo with a skill directory
// and returns the repo path and the HEAD commit SHA.
func createLocalGitSkillRepo(t *testing.T) (repoPath, commitSHA string) {
	t.Helper()
	dir := t.TempDir()

	mustGit := func(args ...string) string {
		t.Helper()
		out, err := exec.Command("git", args...).CombinedOutput()
		if err != nil {
			t.Fatalf("git %v: %v\n%s", args, err, out)
		}
		return strings.TrimSpace(string(out))
	}

	mustGitIn := func(dir string, args ...string) string {
		t.Helper()
		cmd := exec.Command("git", args...)
		cmd.Dir = dir
		out, err := cmd.CombinedOutput()
		if err != nil {
			t.Fatalf("git %v in %s: %v\n%s", args, dir, err, out)
		}
		return strings.TrimSpace(string(out))
	}

	_ = mustGit

	mustGitIn(dir, "init")
	mustGitIn(dir, "config", "user.email", "test@example.com")
	mustGitIn(dir, "config", "user.name", "Test")

	skillMD := filepath.Join(dir, "SKILL.md")
	if err := os.WriteFile(skillMD, []byte("---\nname: git-skill\n---\n# Git Skill\n"), 0o644); err != nil {
		t.Fatalf("write SKILL.md: %v", err)
	}

	mustGitIn(dir, "add", ".")
	mustGitIn(dir, "commit", "-m", "add skill")

	sha := mustGitIn(dir, "rev-parse", "HEAD")
	return dir, sha
}

// repoFixtureDir returns the absolute path to a fixture directory.
func repoFixtureDir(t *testing.T, name string) string {
	t.Helper()
	wd, err := os.Getwd()
	if err != nil {
		t.Fatalf("getwd: %v", err)
	}
	p := filepath.Join(wd, "fixtures", name)
	if _, err := os.Stat(p); err != nil {
		t.Fatalf("fixture dir %s missing: %v", name, err)
	}
	return p
}

File: apps/ponte/tests/e2e/skill.test.ts
import { describe, expect, it } from "bun:test";
import { mkdir, writeFile } from "node:fs/promises";
import { tmpdir as osTmpdir } from "node:os";
import { join } from "node:path";
import { $ } from "bun";
import type { Home } from "./harness";
import { newHarness, skillDoc, sourceEntry } from "./harness";

const isWindows = () => process.platform === "win32";

const addSkill = (h: Home, source: string): Promise<void> =>
  h.appendConfig(sourceEntry("skills", source));

const addGitSkill = (h: Home, url: string, ref: string): Promise<void> =>
  h.appendConfig(sourceEntry("skills", url, ref));

const createLocalGitSkillRepo = async (): Promise<{
  repoPath: string;
  commitSHA: string;
}> => {
  const repoPath = join(
    osTmpdir(),
    `ponte-git-skill-${Math.random().toString(36).slice(2, 8)}`,
  );
  await mkdir(repoPath, { recursive: true });

  const git = async (...args: string[]): Promise<string> => {
    const res = await $`git ${args}`.cwd(repoPath).quiet();
    if (res.exitCode !== 0) {
      throw new Error(
        `git ${args.join(" ")} failed: ${res.stdout}${res.stderr}`,
      );
    }
    return res.stdout.toString();
  };

  await git("init");
  await git("config", "user.email", "test@example.com");
  await git("config", "user.name", "Test");
  await writeFile(join(repoPath, "SKILL.md"), skillDoc("git-skill"));
  await git("add", ".");
  await git("commit", "-m", "add skill");
  const commitSHA = (await git("rev-parse", "HEAD")).trim();

  return { repoPath, commitSHA };
};

describe("skill sync", () => {
  it("appears in every vendor skills dir after sync", async () => {
    if (isWindows()) {
      return; // symlink assertions require Unix
    }
    const h = await newHarness();
    await h.bootstrap();

    await addSkill(h, h.fixture("simple_skill"));

    await h.mustRun("sync");

    for (const [, skillsDir] of Object.entries(h.vendorSkillsDirs())) {
      const skillMD = join(skillsDir as string, "simple-skill", "SKILL.md");
      const got = await h.readFileText(skillMD);
      expect(got).toContain("simple-skill");
    }
    await h.close();
  });

  it("names the link after the SKILL.md name, not the directory", async () => {
    if (isWindows()) {
      return;
    }
    const h = await newHarness();
    await h.bootstrap();

    const directory = join(h.home, "sources", "some-folder");
    await h.writeSkill(directory, skillDoc("declared-name"));
    await addSkill(h, directory);

    await h.mustRun("sync");

    await h.assertSymlinkTo(
      h.vendorSkillPath("claude-code", "declared-name"),
      directory,
    );
    await h.assertMissing(h.vendorSkillPath("claude-code", "some-folder"));
    await h.close();
  });

  it("symlinks the skill straight to its source directory", async () => {
    if (isWindows()) {
      return;
    }
    const h = await newHarness();
    await h.bootstrap();

    const skillFixtureDir = h.fixture("simple_skill");
    await addSkill(h, skillFixtureDir);

    await h.mustRun("sync");

    await h.assertSymlinkTo(
      h.vendorSkillPath("claude-code", "simple-skill"),
      skillFixtureDir,
    );
    await h.close();
  });

  it("symlinks the instruction file to the configured prompt", async () => {
    if (isWindows()) {
      return;
    }
    const h = await newHarness();
    await h.bootstrap();

    await h.assertSymlinkTo(
      h.vendorPaths()["claude-code"],
      h.configPath("AGENTS.md"),
    );
    await h.close();
  });

  it("removes the link when a skill leaves the config", async () => {
    if (isWindows()) {
      return;
    }
    const h = await newHarness();
    await h.bootstrap();

    const before = await h.readFileText(h.configPath("config.toml"));
    await addSkill(h, h.fixture("simple_skill"));
    await h.mustRun("sync");
    await h.assertSymlinkTo(
      h.vendorSkillPath("claude-code", "simple-skill"),
      h.fixture("simple_skill"),
    );

    await h.writeFile(h.configPath("config.toml"), before);
    const { stdout } = await h.mustRun("sync");

    expect(stdout).toContain("stale link(s) removed");
    await h.assertMissing(h.vendorSkillPath("claude-code", "simple-skill"));
    await h.close();
  });

  it("leaves a directory it did not create alone", async () => {
    if (isWindows()) {
      return;
    }
    const h = await newHarness();
    await h.bootstrap();

    const mine = h.vendorSkillPath("claude-code", "mine");
    await h.writeFile(join(mine, "SKILL.md"), "hand written");
    await h.mustRun("sync");

    expect(await h.readFileText(join(mine, "SKILL.md"))).toBe("hand written");
    await h.close();
  });

  it("clones and links a git skill under its declared name", async () => {
    if (isWindows()) {
      return;
    }
    const h = await newHarness();
    await h.bootstrap();

    const { repoPath, commitSHA } = await createLocalGitSkillRepo();
    await addGitSkill(h, `file://${repoPath}`, commitSHA);

    await h.mustRun("sync");

    const skillMD = join(
      h.vendorSkillPath("claude-code", "git-skill"),
      "SKILL.md",
    );
    expect(await h.readFileText(skillMD)).toContain("git-skill");
    await h.close();
  });
});

describe("skill name errors", () => {
  it("rejects a source with no SKILL.md", async () => {
    const h = await newHarness();
    await h.bootstrap();

    const directory = join(h.home, "sources", "empty");
    await h.writeFile(join(directory, "README.md"), "no skill here\n");
    await addSkill(h, directory);

    const { stderr, exitCode } = await h.run("sync");

    expect(exitCode).not.toBe(0);
    expect(stderr).toContain("SKILL.md");
    expect(stderr).toContain(directory);
    await h.close();
  });

  it("rejects a SKILL.md with no frontmatter", async () => {
    const h = await newHarness();
    await h.bootstrap();

    const directory = join(h.home, "sources", "bare");
    await h.writeSkill(directory, "# No frontmatter\n");
    await addSkill(h, directory);

    const { stderr, exitCode } = await h.run("sync");

    expect(exitCode).not.toBe(0);
    expect(stderr).toContain("no frontmatter");
    await h.close();
  });

  it("rejects frontmatter with no name", async () => {
    const h = await newHarness();
    await h.bootstrap();

    const directory = join(h.home, "sources", "nameless");
    await h.writeSkill(directory, "---\ndescription: none\n---\n");
    await addSkill(h, directory);

    const { stderr, exitCode } = await h.run("sync");

    expect(exitCode).not.toBe(0);
    expect(stderr).toContain("declares no name");
    await h.close();
  });

  it("rejects an invalid name", async () => {
    const h = await newHarness();
    await h.bootstrap();

    const directory = join(h.home, "sources", "shouty");
    await h.writeSkill(directory, skillDoc("Not Valid"));
    await addSkill(h, directory);

    const { stderr, exitCode } = await h.run("sync");

    expect(exitCode).not.toBe(0);
    expect(stderr).toContain("Not Valid");
    await h.close();
  });

  it("rejects two sources that declare the same name", async () => {
    const h = await newHarness();
    await h.bootstrap();

    const first = join(h.home, "sources", "one");
    const second = join(h.home, "sources", "two");
    await h.writeSkill(first, skillDoc("twin"));
    await h.writeSkill(second, skillDoc("twin"));
    await addSkill(h, first);
    await addSkill(h, second);

    const { stderr, exitCode } = await h.run("sync");

    expect(exitCode).not.toBe(0);
    expect(stderr).toContain("twin");
    expect(stderr).toContain(second);
    await h.close();
  });
});
